File: src/serial.rs
use arduino_hal::{hal::Atmega, usart::UsartOps, Usart};
use embedded_hal::serial::{Read, Write};
use void::ResultVoidExt;

use crate::millis::millis;

/// The types of sendable messages.
/// Keep up to date with /src/serial.rs
///
/// Note: all messages are big-endian
pub enum SerialMsg {
    /// Just send HandshakeResponse back if you get this.
    ///
    /// Parameter: mode
    ///   0x0000 - let the other party decide
    ///   0x0001 - sync
    ///   0x0002 - other party is slave
    ///   0x0003 - other party is master
    ///
    /// 0xc0
    Handshake { mode: u32 },
    /// Yay, Handshake was successful!
    ///
    /// Parameter: selected_mode
    ///   0x0000 - mode not supported
    ///   0x0001 - sync
    ///   0x0002 - we're slave
    ///   0x0003 - we're master
    ///
    /// 0xc1
    HandshakeResponse { selected_mode: u32 },
    /// P1 is now counting down, and P2 has the specified # of ms.
    ///
    /// 0xc2
    StartP1 { p2_time: u32 },
    /// P2 is now counting down, and P1 has the specified # of ms.
    ///
    /// 0xc3
    StartP2 { p1_time: u32 },
    /// Syncing time. P1 first, then P2, in ms.
    ///
    /// 0xc4
    Sync { p1_time: u32, p2_time: u32 },
    /// Pause both times. The currently running clock finished at the specified
    /// ms.
    ///
    /// 0xc5
    Pause { time: u32 },
    /// P1 won by time.
    ///
    /// 0xc6
    P1Finish,
    /// P2 won by time.
    ///
    /// 0xc7
    P2Finish,
}

impl SerialMsg {
    fn to_u8(&self) -> u8 {
        match *self {
            SerialMsg::Handshake { mode: _ } => 0xc0,
            SerialMsg::HandshakeResponse { selected_mode: _ } => 0xc1,
            SerialMsg::StartP1 { p2_time: _ } => 0xc2,
            SerialMsg::StartP2 { p1_time: _ } => 0xc3,
            SerialMsg::Sync {
                p1_time: _,
                p2_time: _,
            } => 0xc4,
            SerialMsg::Pause { time: _ } => 0xc5,
            SerialMsg::P1Finish => 0xc6,
            SerialMsg::P2Finish => 0xc7,
        }
    }

    fn is_connection_message(&self) -> bool {
        match *self {
            SerialMsg::Handshake { mode: _ } => true,
            SerialMsg::HandshakeResponse { selected_mode: _ } => true,
            _ => false,
        }
    }
}

pub struct SerialHandler<USART: UsartOps<Atmega, RX, TX>, RX, TX> {
    serial: Usart<USART, RX, TX>,
    wait_start: Option<u32>,
    pub connected: bool,
}

/// Handles serial communication between the firmware and website.
/// Keep up to date with /www/src/serial.ts
impl<USART: UsartOps<Atmega, RX, TX>, RX, TX> SerialHandler<USART, RX, TX> {
    fn write_u32(&mut self, v: u32) {
        let bytes = v.to_be_bytes();
        for byte in bytes {
            nb::block!(self.serial.write(byte)).void_unwrap();
        }
    }

    pub fn write(&mut self, msg: SerialMsg) {
        // don't write to serial if not connected
        if self.connected || msg.is_connection_message() {
            self.serial.write_byte(msg.to_u8());
            match msg {
                SerialMsg::Handshake { mode } => {
                    self.write_u32(mode);
                }
                SerialMsg::HandshakeResponse { selected_mode } => {
                    self.write_u32(selected_mode);
                }
                SerialMsg::StartP1 { p2_time } => {
                    self.write_u32(p2_time);
                }
                SerialMsg::StartP2 { p1_time } => {
                    self.write_u32(p1_time);
                }
                SerialMsg::Sync { p1_time, p2_time } => {
                    self.write_u32(p1_time);
                    self.write_u32(p2_time);
                }
                SerialMsg::Pause { time } => {
                    self.write_u32(time);
                }
                SerialMsg::P1Finish => {}
                SerialMsg::P2Finish => {}
            }
        }
    }

    fn read_u32(&mut self) -> u32 {
        ((nb::block!(self.serial.read()).void_unwrap() as u32) << 24)
            | ((nb::block!(self.serial.read()).void_unwrap() as u32) << 16)
            | ((nb::block!(self.serial.read()).void_unwrap() as u32) << 8)
            | nb::block!(self.serial.read()).void_unwrap() as u32
    }

    /// Reads a raw message from the wire and blocks until it's completely received.
    fn raw_read(&mut self) -> nb::Result<SerialMsg, void::Void> {
        let msg = self.serial.read()?;
        match msg {
            0xc0 => Ok(SerialMsg::Handshake {
                mode: self.read_u32(),
            }),
            0xc1 => Ok(SerialMsg::HandshakeResponse {
                selected_mode: self.read_u32(),
            }),
            0xc2 => Ok(SerialMsg::StartP1 {
                p2_time: self.read_u32(),
            }),
            0xc3 => Ok(SerialMsg::StartP2 {
                p1_time: self.read_u32(),
            }),
            0xc4 => Ok(SerialMsg::Sync {
                p1_time: self.read_u32(),
                p2_time: self.read_u32(),
            }),
            0xc5 => Ok(SerialMsg::Pause {
                time: self.read_u32(),
            }),
            _ => {
                // Huh? Malformed message, this isn't good, ignore the message
                Err(nb::Error::WouldBlock)
            }
        }
    }

    pub fn read(&mut self) -> nb::Result<SerialMsg, void::Void> {
        let msg = self.raw_read()?;
        match msg {
            SerialMsg::Handshake { mode } => {
                self.connected = true;
                let selected_mode = match mode {
                    0x0000 => 0x0002, // if we decide, make them a slave
                    0x0001 => 0x0001, // syncing
                    0x0002 => 0x0000, // if we're a slave, it's not supported
                    0x0003 => 0x0003, // we're master
                    _ => 0x0000,      // unsupported value
                };
                self.write(SerialMsg::HandshakeResponse { selected_mode });
                Err(nb::Error::WouldBlock)
            }
            SerialMsg::HandshakeResponse { selected_mode: _ } => {
                // hopefully selected_mode is ok. we don't have the resources to check
                self.connected = true;
                Err(nb::Error::WouldBlock)
            }
            _ => Ok(msg),
        }
    }

    pub fn check_connection(&mut self, timeout_ms: u16) -> nb::Result<bool, void::Void> {
        if let None = self.wait_start {
            self.connected = false;
            self.wait_start = Some(millis());
            self.write(SerialMsg::Handshake { mode: 0x0002 });
        }
        let wait_start = self.wait_start.unwrap();
        match self.read() {
            Ok(_) => {
                // ignore. hopefully result isn't anything important.
                Err(nb::Error::WouldBlock)
            }
            Err(nb::Error::WouldBlock) => {
                if self.connected {
                    Ok(true)
                } else if (millis() - wait_start) >= timeout_ms.into() {
                    Ok(false)
                } else {
                    // ignore. still waiting.
                    Err(nb::Error::WouldBlock)
                }
            }
            Err(x) => Err(x),
        }
    }

    pub fn new(serial: Usart<USART, RX, TX>) -> Self {
        let new = Self {
            serial,
            wait_start: None,
            connected: false,
        };
        new
    }
}

File: www/src/serial.ts
/**
 * The types of sendable messages.
 * Keep up to date with /src/serial.rs
 *
 * Note: all messages are big-endian
 */
const messages = Object.freeze({
  handshake: {
    hex: 0xc0,
    arguments: 1,
  },
  handshakeResponse: {
    hex: 0xc1,
    arguments: 1,
  },
  startP1: {
    hex: 0xc2,
    arguments: 1,
  },
  startP2: {
    hex: 0xc3,
    arguments: 1,
  },
  sync: {
    hex: 0xc4,
    arguments: 2,
  },
  pause: {
    hex: 0xc5,
    arguments: 1,
  },
  p1Finish: {
    hex: 0xc6,
    arguments: 0,
  },
  p2Finish: {
    hex: 0xc7,
    arguments: 0,
  },
} satisfies Record<string, { hex: number; arguments: number }>);

/**
 * Handles serial communication between the firmware and website.
 * Keep up to date with /src/serial.rs
 */
export class SerialHandler {
  connected = false;

  constructor(private port: SerialPort) {}

  async write(message: keyof typeof messages, args: number[] = []) {
    const writer = this.port.writable?.getWriter();
    if (!writer) throw new Error("failed to lock serial writer");
    await writer.ready;
    if (args.length !== messages[message].arguments) {
      throw new Error("provided arguments does not match spec");
    }
    const buffer = new ArrayBuffer(
      1 + // the msg header
        args.length * 4 // one u32 (4 bytes) per argument
    );
    const dv = new DataView(buffer, 0, buffer.byteLength);
    dv.setUint8(0, messages[message].hex);
    args.forEach((arg, i) => {
      dv.setUint32(
        1 + i * 4,
        arg,
        false // big-endian
      );
    });
    await writer.write(new Uint8Array(buffer));
    writer.releaseLock();
    console.info(`wrote ${message}:`, args);
  }

  private async rawRead(): Promise<[keyof typeof messages, number[]]> {
    const read = async (): Promise<[keyof typeof messages, number[]]> => {
      const reader = this.port.readable?.getReader({
        mode: "byob",
      });
      if (!reader) throw new Error("failed to lock serial writer");
      const { value: headerValue } = await reader.read(
        new Uint8Array(new ArrayBuffer(1), 0, 1)
      );
      if (!headerValue) throw new Error("stream ended early");
      const headerDv = new DataView(headerValue.buffer);
      const msgId = headerDv.getUint8(0);
      const message = Object.entries(messages).find(
        ([_, { hex }]) => hex === msgId
      )?.[0] as keyof typeof messages | undefined;
      if (message === undefined) {
        console.warn("unknown serial message header:", msgId);
        // wait for a little, flush reader for some amount of bytes then try again
        await new Promise((r) => setTimeout(r, 100));
        await reader.read(new Uint8Array(256));
        reader.releaseLock();
        return await read();
      }
      if (messages[message].arguments > 0) {
        // Wait for a little to let the clock finish sending the bytes
        await new Promise((r) => setTimeout(r, 100));
        const argsLength = messages[message].arguments * 4;
        const argumentsBuffer = new ArrayBuffer(argsLength);
        const { value: argumentsValue } = await reader.read(
          new Uint8Array(argumentsBuffer, 0, argsLength)
        );
        if (!argumentsValue) throw new Error("stream ended early");
        const argumentsDv = new DataView(argumentsValue.buffer);
        const args: number[] = [];
        for (let offset = 0; offset < argsLength; offset += 4) {
          args.push(
            argumentsDv.getUint32(
              offset,
              false // big-endian
            )
          );
        }
        reader.releaseLock();
        console.info(`read ${message}:`, args);
        return [message, args];
      } else {
        reader.releaseLock();
        console.info(`read ${message}`);
        return [message, []];
      }
    };
    return await read();
  }

  async read(): Promise<[keyof typeof messages, number[]]> {
    const read = async (): Promise<[keyof typeof messages, number[]]> => {
      const [msg, args] = await this.rawRead();
      if (msg === "handshake") {
        this.connected = true;
        let selectedMode = 0x0000;
        switch (args[0]) {
          case 0x0000: {
            // We decide
            selectedMode = 0x0003; // make other party master
            break;
          }
          case 0x0001: {
            // Sync
            selectedMode = 0x0000; // unsupported
            break;
          }
          case 0x0002: {
            // We're slave
            selectedMode = 0x0002; // confirmed
            break;
          }
          case 0x0003: {
            // We're master
            selectedMode = 0x0000; // Unsupported
            break;
          }
          default: {
            selectedMode = 0x0000; // Unsupported
            break;
          }
        }
        this.write("handshakeResponse", [selectedMode]);
        // retry
        return await read();
      } else if (msg === "handshakeResponse") {
        this.connected = true;
        if (args[0] === 0x0000) {
          // they said our mode is unsupported
          console.error("can't negotiate mode");
        }
        // retry
        return await read();
      } else {
        return [msg, args];
      }
    };
    return await read();
  }

  async close() {
    this.port.close();
  }

  async open(baudRate: number) {
    this.port.open({
      baudRate,
    });
  }
}

/**
 * SerialHandler made more JavaScripty (since the original was written in Rust)
 */
export class JsSerialHandler extends SerialHandler {
  onNewMessage:
    | ((message: keyof typeof messages, args: number[]) => void)
    | undefined;

  constructor(port: SerialPort) {
    super(port);

    const readLoop = async () => {
      while (true) {
        const [message, args] = await this.read();
        this.onNewMessage?.call(this, message, args);
      }
    };
    readLoop();
  }
}
